File: emulator/memory/memory.go
/*
Copyright (c) 2019-2021 Andreas T Jonsson

This software is provided 'as-is', without any express or implied
warranty. In no event will the authors be held liable for any damages
arising from the use of this software.

Permission is granted to anyone to use this software for any purpose,
including commercial applications, and to alter it and redistribute it
freely, subject to the following restrictions:

1. The origin of this software must not be misrepresented; you must not
   claim that you wrote the original software. If you use this software
   in a product, an acknowledgment in the product documentation would be
   appreciated but is not required.
2. Altered source versions must be plainly marked as such, and must not be
   misrepresented as being the original software.
3. This notice may not be removed or altered from any source distribution.
*/

package memory

import (
	"fmt"
	"log"
)

type Address uint32

func NewAddress(seg, offset uint16) Address {
	return (Address(seg) << 16) | Address(offset)
}

func (a Address) String() string {
	return fmt.Sprintf("0x%X:0x%X", a.Segment(), a.Offset())
}

func (a Address) Segment() uint16 {
	return uint16(a >> 16)
}

func (a Address) Offset() uint16 {
	return uint16(a & 0xFFFF)
}

func (a Address) Pointer() Pointer {
	return NewPointer(a.Segment(), a.Offset())
}

func (a Address) AddInt(i int) Address {
	return (Address(a) & 0xFFFF0000) | Address(a.Offset()+uint16(i))
}

type Pointer uint32

func NewPointer(seg, offset uint16) Pointer {
	return (Pointer(seg)*0x10 + Pointer(offset)) & 0xFFFFF
}

func (p Pointer) String() string {
	return fmt.Sprintf("0x%X", uint32(p))
}

type Memory interface {
	ReadByte(addr Pointer) byte
	WriteByte(addr Pointer, data byte)
}

type IO interface {
	In(port uint16) byte
	Out(port uint16, data byte)
}

type DummyIO struct{}

func (m *DummyIO) In(port uint16) byte {
	log.Printf("reading unmapped IO port: 0x%X", port)
	return 0xFF
}

func (m *DummyIO) Out(port uint16, data byte) {
	log.Printf("writing unmapped IO port: 0x%X", port)
}

type DummyMemory struct{}

func (m *DummyMemory) ReadByte(addr Pointer) byte {
	log.Printf("reading unmapped memory: 0x%X", addr)
	return 0xFF
}

func (m *DummyMemory) WriteByte(addr Pointer, data byte) {
	log.Printf("writing unmapped memory: 0x%X", addr)
}

File: lib/vxt/memory.c
// Copyright (c) 2019-2024 Andreas T Jonsson <mail@andreasjonsson.se>
//
// This software is provided 'as-is', without any express or implied
// warranty. In no event will the authors be held liable for any damages
// arising from the use of this software.
//
// Permission is granted to anyone to use this software for any purpose,
// including commercial applications, and to alter it and redistribute it
// freely, subject to the following restrictions:
//
// 1. The origin of this software must not be misrepresented; you must not
//    claim that you wrote the original software. If you use this software in
//    a product, an acknowledgment (see the following) in the product
//    documentation is required.
//
//    This product make use of the VirtualXT software emulator.
//    Visit https://virtualxt.org for more information.
//
// 2. Altered source versions must be plainly marked as such, and must not be
//    misrepresented as being the original software.
//
// 3. This notice may not be removed or altered from any source distribution.

#include "common.h"
#include <vxt/vxtu.h>

struct memory {
    vxt_pointer base;
    bool read_only;
    int size;
    vxt_byte data[];
};

static vxt_byte read(struct memory *m, vxt_pointer addr) {
    ENSURE((int)(addr - m->base) < m->size);
    return m->data[addr - m->base];
}

static void write(struct memory *m, vxt_pointer addr, vxt_byte data) {
    ENSURE((int)(addr - m->base) < m->size);
    if (!m->read_only) {
        m->data[addr - m->base] = data;
    } else {
        VXT_LOG("writing to read-only memory: [0x%X] = 0x%X", addr, data);
    }
}

static vxt_error install(struct memory *m, vxt_system *s) {
    vxt_system_install_mem(s, VXT_GET_PERIPHERAL(m), m->base, (m->base + (vxt_pointer)m->size) - 1);
    return VXT_NO_ERROR;
}

static const char *name(struct memory *m) {
    return m->read_only ? "ROM" : "RAM";
}

VXT_API struct vxt_peripheral *vxtu_memory_create(vxt_allocator *alloc, vxt_pointer base, int amount, bool read_only) {
    int size = VXT_PERIPHERAL_SIZE(memory) + amount;
    struct VXT_PERIPHERAL(struct memory) *PERIPHERAL;
    *(void**)&PERIPHERAL = alloc(NULL, size);

    vxt_memclear(PERIPHERAL, size);
    struct memory *mem = VXT_GET_DEVICE(memory, PERIPHERAL);

    #ifndef VXTU_MEMCLEAR
        if (!read_only) vxtu_randomize(mem->data, amount, (intptr_t)PERIPHERAL);
    #endif

    mem->base = base;
    mem->read_only = read_only;
    mem->size = amount;

    PERIPHERAL->install = &install;
    PERIPHERAL->name = &name;
    PERIPHERAL->io.read = &read;
    PERIPHERAL->io.write = &write;

    return (struct vxt_peripheral*)PERIPHERAL;
}

VXT_API void *vxtu_memory_internal_pointer(struct vxt_peripheral *p) {
    return VXT_GET_DEVICE(memory, p)->data;
}

VXT_API bool vxtu_memory_device_fill(struct vxt_peripheral *p, const vxt_byte *data, int size) {
    struct memory *m = VXT_GET_DEVICE(memory, p);
    ENSURE(data);
    if (m->size < size)
        return false;
    for (int i = 0; i < size; i++)
        m->data[i] = data[i];
    return true;
}
